### PR TITLE
Convert java.util.Date to java.time.Instant (jsr-310)

### LINKED ISSUE #57: [Performance] Improve parsing dates and times
Actually, `io.vertx.httpproxy.impl.ParseUtils` use SimpleDateFormat to parse a date. 

It would certainly be possible to improve memory consumption, perhaps using another more efficient method (LocalDateTime.parse ?)

**Java Flight Recorder during 300s**
I observe more than 4% of the memory consumption of my application linked to this analysis.

Screenshot showing 1,3% of memory usage.
image

Observed with the version 4.5.1 (from quarkus 3.7.0)

### PR DESCRIPTION
**Pull request by lucaspouzac**
Convert JSR-310 for Fix #57

### CODE REVIEW

**Comment by lucaspouzac** [comment]
I create ParseUtilsTest before migration. 

Must be reread :)

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java [hunk 1]
```diff
@@ -55,30 +66,12 @@ public static Date parseWarningHeaderDate(String value) {
...
-    return format;
-  }
-
-  public static String formatHttpDate(Date date) {
-    return RFC_1123_DATE_TIME().format(date);
+  public static String formatHttpDate(Instant date) {
+    return DateTimeFormatter.RFC_1123_DATE_TIME.format(OffsetDateTime.ofInstant(date, ZoneOffset.UTC));
   }
 
   // http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.3.1
```
Comment (tsegismont) [inline src/main/java/io/vertx/httpproxy/impl/ParseUtils.java:73]:
```suggestion
  // https://www.rfc-editor.org/rfc/rfc9110#http.date
```

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java [hunk 2]
```diff
@@ -87,17 +80,17 @@ public static Date parseHttpDate(String value) throws Exception {
...
           }
           return null;
         }  else if (c == ' ') {
           String s = value.substring(0, sep);
           if (parseWkday(s) != null) {
             // asctime-date
-            return ASC_TIME().parse(value);
+            return Instant.from(ASC_TIME.parse(value));
           }
           return null;
```
Comment (tsegismont) [inline src/main/java/io/vertx/httpproxy/impl/ParseUtils.java:95]: I think we could simplify this method to:

```java
  public static Instant parseHttpDate(String value) throws Exception {
    int pos = value.indexOf(',');
    if (pos == 3) { // e.g. Sun, 06 Nov 1994 08:49:37 GMT
      return RFC_1123_DATE_TIME.parse(value, Instant::from);
    }
    if (pos == -1) { // e.g. Sun Nov  6 08:49:37 1994
      return ANSI_C_ASCTIME.parse(value, Instant::from);
    }
    return RFC_850_DATE_TIME.parse(value, Instant::from); // e.g. Sunday, 06-Nov-94 08:49:37 GMT
  }
```

See https://www.rfc-editor.org/rfc/rfc9110.html#name-date-time-formats

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java [hunk 3]
```diff
@@ -10,26 +10,37 @@
...
 /**
  * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
  */
 public class ParseUtils {
 
-  public static Date parseHeaderDate(String value) {
+  public static final DateTimeFormatter RFC_850_DATE_TIME = new DateTimeFormatterBuilder()
+    .appendPattern("EEEE, dd-MMM-yy HH:mm:ss")
+    .parseLenient()
+    .appendLiteral(" UTC")
```
Comment (tsegismont) [inline src/main/java/io/vertx/httpproxy/impl/ParseUtils.java:26]:
```suggestion
    .appendLiteral(" GMT")
```
See https://www.rfc-editor.org/rfc/rfc9110.html#name-date-time-formats and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java [hunk 4]
```diff
@@ -10,26 +10,37 @@
...
+    .toFormatter(Locale.US).withZone(ZoneId.of("UTC"));
```
Comment (tsegismont) [inline src/main/java/io/vertx/httpproxy/impl/ParseUtils.java:27]:
```suggestion
    .toFormatter(Locale.US)
    .withZone(ZoneOffset.UTC);
```

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java [hunk 5]
```diff
@@ -10,26 +10,37 @@
...
+
+  public static final DateTimeFormatter ASC_TIME = new DateTimeFormatterBuilder()
+    .appendPattern("EEE MMM d HH:mm:ss yyyy")
+    .parseLenient()
+    .toFormatter(Locale.US).withZone(ZoneId.of("UTC"));
```
Comment (tsegismont) [inline src/main/java/io/vertx/httpproxy/impl/ParseUtils.java:32]:
```suggestion
    .toFormatter(Locale.US)
    .withZone(ZoneOffset.UTC);
```

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java (hunk 2 above)
Comment (tsegismont) [reply src/main/java/io/vertx/httpproxy/impl/ParseUtils.java:95]: And then you can get rid of the two methods below

**Review by tsegismont — COMMENTED** [review]
Thank you @lucaspouzac , I did a first pass to your PR, good job

File: src/main/java/io/vertx/httpproxy/impl/ParseUtils.java (hunk 1 above)
Comment (lucaspouzac) [reply src/main/java/io/vertx/httpproxy/impl/ParseUtils.java]: Done

File: src/test/java/io/vertx/httpproxy/impl/ParseUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import io.vertx.ext.unit.TestContext;
+import io.vertx.httpproxy.TestBase;
+import org.junit.Test;
+
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+
+import static junit.framework.TestCase.assertEquals;
+
+public class ParseUtilsTest extends TestBase {
```
Comment (tsegismont) [inline src/test/java/io/vertx/httpproxy/impl/ParseUtilsTest.java:12]:
```suggestion
public class ParseUtilsTest {
```

It seems there is no need to create an instance of Vertx before each test

**Review by tsegismont — APPROVED** [review]
Thank you @lucaspouzac 

LGTM, if a minor comment

**Comment by tsegismont** [comment]
Can you revert your last commit please? You shouldn't format files which were not modified

File: src/test/java/io/vertx/httpproxy/impl/ParseUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+
+import static junit.framework.TestCase.assertEquals;
+
+@RunWith(VertxUnitRunner.class)
```
Comment (tsegismont) [inline src/test/java/io/vertx/httpproxy/impl/ParseUtilsTest.java:13]:
```suggestion
```

Why do you need the extension? It seems to me the test is pretty straightforward: it uses date time formats and that's it.

**Comment by tsegismont** [comment]
Thank you

**Comment by vietj** [comment]
oh nice, thanks @lucaspouzac !

**Pull request merged**